tests/core/test_decorator.py

```diff
@@ -0,0 +1,31 @@
+
+import pytest
+from fastats.core.decorator import fs
+
+
+@pytest.mark.xfail
+def test_decorated_func_kwargs():
+    @fs
+    def square(x):
+        return x ** 2
+
+    @fs
+    def func_that_takes_func_kwarg(arg, func=abs):
```

Comment:
At the moment this API is not going to work - the reason for the `@fs` decorator is to allow functions to be passed as kwargs to a numba jit function, which currently numba doesn't support, so this signature will fail to jit. 

However, your example here is far more interesting - I hadn't thought about defining kwarg substitutions in the `@fs`-decorated signature before, but this actually ties in nicely with a comment that @pawroman made to me recently - that we need a way to define what the replaceable / substitutable functions are.

This could allow auto-documentation of which functions are able to be replaced using the AST substitution, without requiring a more complicated decorator signature.

We could easily use this commit as an API request for how to define which items in the function are actually able to be replaced by the `@fs` decorator, so effectively instead of:

```python
@fs
def calc(x):
    y = square(x)
    z = cube(y)
    return z / 2
```
which would allow any caller to replace either `square` or `cube`, we could have a way of defining that only `square` is replaceable using:

```python
@fs
def calc(x, square=square):
    y = square(x)
    z = cube(y)
    return z / 2
```
which would stop `cube` from being replaced by any callers.

@fastats/core / @AaronCritchley - thoughts?

Reply:
```python
@fs
def calc(x, square=square):
    y = square(x)
    z = cube(y)
    return z / 2
```

This requires some AST pre-processing to work out that `square` is callable and should be substituted. But I like the idea of making things explicit and possibly not injecting anything ever unless it's an explicit parameter. Magical injection is nice to have in some cases, but IMO having it on by default, we violate a bunch of Python guidelines and general sanity. It would be very nice to have a context manager that says "magical injection happens here", like:

```python
@fs
def calc(x, square=square):
    y = square(x)
    z = cube(y)
    return z / 2

# fails because no argument in calc - Python semantics preserved by default
calc(cube=lambda x: x ** 4)

# works - magical injection is scoped and more explicit
with auto_inject():
   calc(cube=lambda x: x ** 4)  # silly example, but you get the idea
```

What do you think?